log,cli: Stop serving earlier when log.Fatal is called

bdarnell opened the pull request:
log.Fatal can take a while (there are 10s timeouts on sending crash
reports and writing to the logs), and during that time we don't want
to serve any bad data (for example, if the fatal error is due to
observed clock inconsistencies).

Fixes #23414

Release note (bug fix): Reduced risk that a node in the process of
crashing could serve inconsistent data.

bdarnell committed:
log,cli: Stop serving earlier when log.Fatal is called

log.Fatal can take a while (there are 10s timeouts on sending crash
reports and writing to the logs), and during that time we don't want
to serve any bad data (for example, if the fatal error is due to
observed clock inconsistencies).

Fixes #23414

Release note (bug fix): Reduced risk that a node in the process of
crashing could serve inconsistent data.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23560)
<!-- Reviewable:end -->

knz commented on pkg/cli/start.go:
```diff
@@ -689,7 +689,21 @@ func runStart(cmd *cobra.Command, args []string) error {
+		// reaching into the server objects and closing all the
+		// connections while they're in use. That would be more in line
+		// with the expected effect of a log.Fatal.
+		stopper.Stop(ctx)
```
This will cause the exit in cobra (which will process the return here) to race with the exit in log.Fatal.

Set `returnErr` here so that both will see the same process exit status (255).

knz commented on pkg/util/log/clog.go:
```diff
@@ -775,6 +783,14 @@ func (l *loggingT) outputLogEntry(s Severity, file string, line int, msg string)
 	var stacks []byte
 	var fatalTrigger chan struct{}
 	if s == Severity_FATAL {
+		fatalChanMu.Lock()
```
The logger is already locked above (l.mu.lock) so I think this lock is unneeded.

bdarnell commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cli/start.go, line 703 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23560#-L719SWbx49FxD7lIiPK-r1-703:-L75PBj0BKee8jSGHwyi:b-ukzuxl) ([raw file](https://github.com/cockroachdb/cockroach/blob/22fab556b64d92421abd1086b08e252c5c349cae/pkg/cli/start.go#L703)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This will cause the exit in cobra (which will process the return here) to race with the exit in log.Fatal.

Set `returnErr` here so that both will see the same process exit status (255).
</blockquote></details>

This doesn't return, it will block forever in the empty select. 

---

*[pkg/util/log/clog.go, line 786 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23560#-L719SWbx49FxD7lIiPL-r1-786:-L75PJ1k0_P_1_xXEVHH:b-fd3b51) ([raw file](https://github.com/cockroachdb/cockroach/blob/22fab556b64d92421abd1086b08e252c5c349cae/pkg/util/log/clog.go#L786)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The logger is already locked above (l.mu.lock) so I think this lock is unneeded.
</blockquote></details>

There are multiple `loggingT` objects. It's a theoretical concern because we'll only have fatal logging on the primary one, but since I introduced the new channel at global scope I think the new mutex is warranted.

(Should the channel be on `loggingT` instead of increasing this package's use of singletons? Maybe, but Fatal is inherently process-scoped)

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23560)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 1 of 2 files at r1.
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cli/start.go, line 703 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23560#-L719SWbx49FxD7lIiPK-r1-703:-L75_OrIEOxIMFuN3k5v:b4wqguw) ([raw file](https://github.com/cockroachdb/cockroach/blob/22fab556b64d92421abd1086b08e252c5c349cae/pkg/cli/start.go#L703)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This doesn't return, it will block forever in the empty select. 
</blockquote></details>

Oh that's what an empty select does. TIL. Thanks

---

*[pkg/util/log/clog.go, line 786 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23560#-L719SWbx49FxD7lIiPL-r1-786:-L75_Vox8GOcPbd2LGcS:bd6jji2) ([raw file](https://github.com/cockroachdb/cockroach/blob/22fab556b64d92421abd1086b08e252c5c349cae/pkg/util/log/clog.go#L786)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

There are multiple `loggingT` objects. It's a theoretical concern because we'll only have fatal logging on the primary one, but since I introduced the new channel at global scope I think the new mutex is warranted.

(Should the channel be on `loggingT` instead of increasing this package's use of singletons? Maybe, but Fatal is inherently process-scoped)
</blockquote></details>

I'd argue the channel should be on loggingT yes. 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23560)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
OK, moved the channel onto `loggingT`.

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/util/log/clog.go, line 790 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23560#-L76D-ULkrI0I21M3VXJ:-L76D-UM98viLLZaYWxd:b-fp6dno) ([raw file](https://github.com/cockroachdb/cockroach/blob/65ced7d8a86fe3e4fb56299cb6bf94027f8859ed/pkg/util/log/clog.go#L790)):*
> ```Go
> 		case <-l.fatalCh:
> 		default:
> 			close(l.fatalCh)
> ```

Naively this isn't correct, right? Two goroutines could run into the default branch and both try to close.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23560#-:-L76DDhRh_GUPPJheN-W:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/util/log/clog.go, line 790 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23560#-L76D-ULkrI0I21M3VXJ:-L76J7epzSk3aSU1ud55:b-4hkh9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/65ced7d8a86fe3e4fb56299cb6bf94027f8859ed/pkg/util/log/clog.go#L790)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Naively this isn't correct, right? Two goroutines could run into the default branch and both try to close.
</blockquote></details>

No, because we're under `l.mu.Lock()`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23560)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/util/log/clog.go, line 790 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23560#-L76D-ULkrI0I21M3VXJ:-L76KWXgaNbWVNuofzuy:bcn3o5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/65ced7d8a86fe3e4fb56299cb6bf94027f8859ed/pkg/util/log/clog.go#L790)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

No, because we're under `l.mu.Lock()`. 
</blockquote></details>

Ah, right.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23560)*
<!-- Sent from Reviewable.io -->

bdarnell committed:
cli: Don't shadow ctx variable

For some reason vet is currently fine with this but starts complaining
with my next change to this file.

Release note: None
